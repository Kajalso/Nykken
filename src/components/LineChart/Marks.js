import { line, curveNatural } from "d3";

export const Marks = ({
  data,
  xScale,
  yScale,
  xValue,
  yValue,
  xFormat,
  circleRadius = 7,
  curveStyle = curveNatural,
}) => (
  <>
    <path
      className="mark-temp-line"
      d={line()
        .x((d) => xScale(xValue(d)))
        .y((d) => yScale(yValue(d)))
        .curve(curveStyle)(data)}
    />
    {data.map((d, i) => (
      <circle
        className="mark-circle"
        key={i}
        cx={xScale(xValue(d))}
        cy={yScale(yValue(d))}
        r={circleRadius}
      >
        <title>{xFormat(xValue(d)) + ": " + yValue(d)}</title>
      </circle>
    ))}
  </>
);
