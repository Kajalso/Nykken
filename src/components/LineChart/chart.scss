$colorBlue: rgb(156, 204, 248);
$colorRed: rgb(250, 143, 143);
$colorPurple: rgb(214, 106, 241);

/* Chart styling */
.chart {
  svg {
    overflow: visible;
  }
}

/* Marks styling */
.mark-circle {
  fill: $colorPurple;
}

.mark-temp-line {
  fill: none;
  stroke: $colorPurple;
  stroke-width: 3;
  stroke-linejoin: round;
  stroke-linecap: round;
}

.axis-label {
  text-anchor: middle;
  font-size: small;
  font-weight: 600;
  fill: #635f5d;
}

.axis-date {
  text-anchor: middle;
  font-size: smaller;
  font-weight: 600;
  fill: #635f5d;
}

/* Axes styling */

.tick-line {
  stroke: #c0c0bb;
}

.tick-text-x {
  text-anchor: middle;
  font-size: x-small;
  fill: #635f5d;
}

.tick-text-y {
  text-anchor: end;
  font-size: x-small;
  fill: #635f5d;
}
