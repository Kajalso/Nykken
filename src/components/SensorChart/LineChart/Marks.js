import { area, line, curveNatural } from "d3";

import { useColors } from "../../../styles/useColors";

export const Marks = ({
  data,
  dataInfo,
  xScale,
  yScale,
  xValue,
  yValue,
  xFormat,
  innerHeight,
  circleRadius = 1,
  curveStyle = curveNatural,
}) => {
  const id = dataInfo.data_identifier;
  const colors = useColors();

  let markColor = colors.blue;

  // Set color to purple for Radiation, Humidity and Wind Speed
  if (id === 3 || id === 4 || id === 6) {
    markColor = colors.purple;
  }

  // Set area fill for Snow Melting Tank and Waste Water
  const areaChart = id === 7 || id === 9 || id === 11;

  return (
    <>
      <path
        className="mark-temp-line"
        stroke={markColor}
        d={line()
          .x((d) => xScale(xValue(d)))
          .y((d) => yScale(yValue(d)))
          .curve(curveStyle)(data)}
      />
      {areaChart && (
        <path
          className="mark-temp-area"
          fill={markColor}
          d={area()
            .x((d) => xScale(xValue(d)))
            .y0(innerHeight)
            .y1((d) => yScale(yValue(d)))
            .curve(curveStyle)(data)}
        />
      )}
      {data.map((d, i) => (
        <circle
          fill={markColor}
          className="mark-circle"
          key={i}
          cx={xScale(xValue(d))}
          cy={yScale(yValue(d))}
          r={circleRadius}
        >
          <title>
            {xFormat(xValue(d)) + ": " + yValue(d) + " " + dataInfo.unit}
          </title>
        </circle>
      ))}
    </>
  );
};
