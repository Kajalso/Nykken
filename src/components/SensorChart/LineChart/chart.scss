$colorBlue: rgb(156, 204, 248);
$colorRed: rgb(250, 143, 143);
$colorPurple: rgb(214, 106, 241);
$colorLine: $colorBlue;

/* Chart styling */
.chart {
  svg {
    overflow: visible;
  }
}

/* Marks styling */
.mark-circle {
  //fill: $colorLine;
}

.mark-temp-line {
  fill: none;
  //stroke: $colorLine;
  stroke-width: 3;
  stroke-linejoin: round;
  stroke-linecap: round;
}

.axis-label {
  font-weight: 400;
  fill: #635f5d;
}

.axis-label.y {
  text-anchor: middle;
}

.axis-date {
  text-anchor: middle;
  font-size: small;
  font-weight: 500;
  fill: #635f5d;
}

/* Axes styling */

.tick-line {
  stroke: #c0c0bb;
}

.tick-text-x {
  text-anchor: middle;
  font-size: x-small;
  fill: #635f5d;
}

.tick-text-y {
  text-anchor: end;
  font-size: x-small;
  fill: #635f5d;
}
