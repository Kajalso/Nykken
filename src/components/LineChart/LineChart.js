import React from "react";
import { scaleTime, scaleLinear, extent, timeFormat, curveMonotoneX } from "d3";

import { AxisBottom } from "./AxisBottom";
import { AxisLeft } from "./AxisLeft";
import { Marks } from "./Marks";

import "./chart.scss";

const width = 700;
const height = 400;
const margin = { top: 10, right: 50, bottom: 50, left: 70 };
const circleRadius = 3;

const innerHeight = height - margin.top - margin.bottom;
const innerWidth = width - margin.right - margin.left;

// X values
const xValue = (d) => new Date(d.time_stamp_utc);
let xAxisLabel = "Time";
const xAxisLabelOffset = 40;

// Y values
const yValue = (d) => +d.measurement;
let yAxisLabel = "Measurement";
const yAxisLabelOffset = 45;

// Axis formats
const xAxisTickFormat = timeFormat("%I:%M");

export const LineChart = ({ data = [], dataInfo = {} }) => {
  if (!data || !dataInfo) {
    return <pre>Loading chart...</pre>;
  }

  // Linear scale for x values
  const xScale = scaleTime()
    .domain(extent(data, xValue)) // extent-function replaces min, max
    .range([0, innerWidth])
    .nice(); // Adjusts the axis to prevent overlap

  // Linear scale for y values
  const yScale = scaleLinear()
    .domain(extent(data, yValue))
    .range([innerHeight, 0])
    .nice();

  return (
    <div className="chart">
      <h4 className="section-title">{dataInfo.description}</h4>
      <div className="data">
        <svg width={width} height={height}>
          <g transform={`translate(${margin.left}, ${margin.top})`}>
            <AxisBottom
              xScale={xScale}
              innerHeight={innerHeight}
              tickFormat={xAxisTickFormat}
              tickOffset={10}
            />
            <AxisLeft yScale={yScale} innerWidth={innerWidth} tickOffset={7} />
            <Marks
              data={data}
              xScale={xScale}
              yScale={yScale}
              xValue={xValue}
              yValue={yValue}
              xFormat={xAxisTickFormat}
              circleRadius={circleRadius}
              curveStyle={curveMonotoneX}
            />
            <text
              x={innerWidth / 2}
              y={innerHeight + xAxisLabelOffset}
              className={"axis-label"}
            >
              {xAxisLabel + " (min)"}
            </text>
            <text
              className={"axis-label"}
              transform={`translate(${-yAxisLabelOffset},
                ${innerHeight / 2}) rotate(-90)`}
            >
              {yAxisLabel + " (" + dataInfo.unit + ")"}
            </text>
          </g>
        </svg>
      </div>
    </div>
  );
};
