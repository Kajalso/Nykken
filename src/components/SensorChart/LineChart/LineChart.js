import React, { useState, useEffect } from "react";
import { scaleTime, scaleLinear, extent, curveMonotoneX } from "d3";

import { AxisBottom } from "./Axes/AxisBottom";
import { AxisLeft } from "./Axes/AxisLeft";
import { Marks } from "./Marks";

import { useChartProps, useGroupProps } from "../../../styles/useChartStyles";

import "../chart.scss";

const circleRadius = 2;

export const LineChart = React.forwardRef(
  ({ data = [], dataInfo = {}, inGroup }, ref) => {
    let {
      width,
      height,
      margin,
      innerHeight,
      innerWidth,
      xValue,
      xAxisLabel,
      xAxisLabelOffset,
      xAxisDateOffset,
      xAxisTickFormat,
      yValue,
      yAxisLabel,
      yAxisLabelOffset,
      dateFormat,
    } = useChartProps();

    /*
    const [chartWidth, setChartWidth] = useState(width);
    let tempChartWidth = width;

     window.addEventListener("resize", () => {
      if (document.body.clientWidth > 1025) {
        console.log("Large screen.");
        tempChartWidth = 800;
      }
      if (document.body.clientWidth > 768) {
        console.log("Medium screen.");
        tempChartWidth = 600;
      } else {
        console.log("Small screen");
        tempChartWidth = 400;
      }
    });

    useEffect(() => {
      setChartWidth(tempChartWidth);
    }, [chartWidth]); */

    // Group props
    let {
      groupWidth,
      groupHeight,
      groupMargin,
      groupInnerHeight,
      groupInnerWidth,
    } = useGroupProps();

    if (inGroup) {
      width = groupWidth;
      height = groupHeight;
      margin = groupMargin;
      innerWidth = groupInnerWidth;
      innerHeight = groupInnerHeight;
    }

    if (!data || !dataInfo) {
      return <pre>Loading chart...</pre>;
    }

    // Linear scale for x values
    const xScale = scaleTime()
      .domain(extent(data, xValue)) // Extent-function replaces min, max
      .range([0, innerWidth])
      .nice(); // Adjusts the axis to prevent overlap

    // Linear scale for y values
    const yScale = scaleLinear()
      .domain(extent(data, yValue))
      .range([innerHeight, 0])
      .nice();

    return (
      <div className="chart" ref={ref}>
        <svg width={width} height={height}>
          <g transform={`translate(${margin.left}, ${margin.top})`}>
            <AxisBottom
              xScale={xScale}
              innerHeight={innerHeight}
              tickFormat={xAxisTickFormat}
              tickOffset={10}
            />
            <AxisLeft yScale={yScale} innerWidth={innerWidth} tickOffset={7} />
            <Marks
              data={data}
              dataInfo={dataInfo}
              xScale={xScale}
              yScale={yScale}
              xValue={xValue}
              yValue={yValue}
              xFormat={xAxisTickFormat}
              circleRadius={circleRadius}
              curveStyle={curveMonotoneX}
              innerHeight={innerHeight}
            />
            <text
              x={innerWidth - yAxisLabelOffset}
              y={innerHeight + xAxisLabelOffset}
              className={"axis-label x small"}
            >
              {xAxisLabel}
            </text>
            <text
              x={innerWidth / 2}
              y={innerHeight + xAxisDateOffset}
              className={"axis-date"}
            >
              {dateFormat(data[0])}
            </text>
            <text
              className={"axis-label y"}
              transform={`translate(${-yAxisLabelOffset},
                ${innerHeight / 2}) rotate(-90)`}
            >
              {yAxisLabel(dataInfo) + " (" + dataInfo.unit + ")"}
            </text>
          </g>
        </svg>
      </div>
    );
  }
);
