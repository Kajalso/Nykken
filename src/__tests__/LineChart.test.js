import React from 'react';
import { render } from '@testing-library/react'
import { LineChart } from '../components/SensorChart/LineChart/LineChart';


//Surpressing warning as suggested by:
const originalError = global.console.error;
beforeAll(() => {
	global.console.error = jest.fn((...args) => {
		if (typeof args[0] === 'string' && args[0].includes('Please upgrade to at least react-dom@16.9.0')) {
	      return
	    }
	    return originalError.call(console, args)
	});
});

afterAll(() => {
  	global.console.error.mockRestore();
});

describe('renders without crashing and takes snapshot', () => {
    test('renders correctly', () => {
        render(< LineChart />);
    });
    test('takes snapshot', () => {
        const { component } = render(< LineChart />);
        expect(component).toMatchSnapshot();
    })
})