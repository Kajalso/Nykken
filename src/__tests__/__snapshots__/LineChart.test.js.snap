// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing and takes snapshot takes snapshot 1`] = `undefined`;
